### ISSUE TITLE
Settings set via builtin nvim client freeze emmylua about half the time

### ISSUE THREAD

**Issue by chrisgrieser** [EmmyLuaLs/emmylua-analyzer-rust#678]
### Bug description
I am using the built-in lsp client on nvim v0.11.3 on macOS, and trying to configure emmylua such as below. 

```lua
-- ~/.config/nvim/lsp/emmylua_ls.lua
return {
	settings = {
		Lua = {
			completion = { postfix = "." },
		},
	},
}
```

While the configuration does take effect (`.` does trigger postfix completions now), I noticed that this freezes emmylua on startup, about half the time. 

### Reproduction conditions
It appears the freeze occurs anytime you add a key to `settings`. With this config, the freeze never occurs:

```lua
-- freeze does not occur anymore
-- ~/.config/nvim/lsp/emmylua_ls.lua
return {
	settings = {
	},
}
```
```lua
-- 50% change of freeze occuring again
-- ~/.config/nvim/lsp/emmylua_ls.lua
return {
	settings = {
		Lua  = {}
	},
}
```

I checked the logs in `~/.local/share/emmylua_ls/logs/`, and there is no error there, whether the freeze occurs or not. There is also no error message during runtime. 

The only thing I could notice was that when the freeze does occur, some client requests seem to be stuck as `pending`: Running `vim.lsp.get_clients({ name = "emmylua_ls" })[1].requests` when the freeze occurs results in this output like this:

```lua
{ vim.NIL, {
    bufnr = 1,
    method = "textDocument/documentSymbol",
    type = "pending"
  }, {
    bufnr = 1,
    method = "textDocument/references",
    type = "pending"
  }, {
    bufnr = 1,
    method = "textDocument/semanticTokens/full",
    type = "pending"
  }, {
    bufnr = 1,
    method = "textDocument/foldingRange",
    type = "pending"
  }, {
    bufnr = 1,
    method = "textDocument/inlayHint",
    type = "pending"
  }, {
    bufnr = 1,
    method = "textDocument/documentHighlight",
    type = "pending"
  }, {
    bufnr = 1,
    method = "textDocument/documentSymbol",
    type = "pending"
  } }
```

Running the same when the freeze does not occur, shows one or two `pending` items at most, or even an empty list.

**Comment by CppCXY** [comment]
Then I suggest you use .emmyrc.json,  the settings in nvim. I'm not sure when those are sent to the server. At startup, I send a request to the LSP client to get all the settings, and if this process is slow, I will wait. Additionally, I support the configuration request; when I receive this request, I reload the workspace. Only the .emmyrc.json configuration method is officially supported by design—other methods I personally do not use.

You can set the log level to debug to verify this.
```
--log-level debug
```

**Comment by chrisgrieser** [comment]
> Only the .emmyrc.json configuration method is officially supported by design—other methods I personally do not use.

That's somewhat impractical: if you have many small lua projects, you have to copy the same config over and over again. Also, project-specific files such as `emmyrc` make sense for project-scoped configs, but sometimes different people working on the same project prefer different configs. For example, one person preferring `.` as their postfix character, while another prefers `@`.

Furthermore, configuring via `settings` is something that works for every LSP I use, only with emmylua there are issues with that.

> Then I suggest you use .emmyrc.json, the settings in nvim.

yeah, it's a bit cumbersome, but that's what I am currently doing now. (Luckily, emmylua has much saner defaults than `lua_ls`, so that it's at least not much config – thanks for that!)

**Comment by CppCXY** [comment]
If you can provide me with a reproducible example, I can try to take a look at it on Monday.

**Comment by chrisgrieser** [comment]
I can reproduce it by opening `test.lua`

```lua
-- ~/.config/nvim/lsp/emmylua_ls.lua
return {
    cmd = { "emmylua_ls" },
    filetypes = { "lua" },
    root_markers = {
    	".luarc.json",
    	".emmyrc.json",
    	".luacheckrc",
    	".git",
    },
    workspace_required = false,
    
    settings = {
    	Lua = {
    		completion = { postfix = "." },
    	},
    },
}
```

```jsonc
// ./.emmyrc.json
{
	"runtime": {
		"version": "LuaJIT"
	}
}
```

```lua
-- ./test.lua
-- should report two diagnostics, for `foo` and for `bar`
local function foo()
  print(bar)
end
```

**Comment by CppCXY** [comment]
can you share the log file？

**Comment by chrisgrieser** [comment]
sure, this is the log on a startup where the freeze occurs.

[Users_chrisgrieser_.config_nvim_debug.log](https://github.com/user-attachments/files/21559438/Users_chrisgrieser_.config_nvim_debug.log)

**Comment by CppCXY** [comment]
it's probably one of the files in here that's causing the loading process to get stuck.

**Comment by CppCXY** [comment]
Can I get the files in `/Users/chrisgrieser/.config/nvim/debug`

**Comment by chrisgrieser** [comment]
It's just `test.lua` and `.emmyrc.json` from above.

I tested it in multiple repos, large and small, and the problem occurs in all of them.

**Comment by CppCXY** [comment]
i download your .config repo, it work well,  I do think it should be related to neovim and some of its plugins. For me, it's too late now, I'll look at it again tomorrow.

**Comment by chrisgrieser** [comment]
I tested in nvim without neovide, with all plugins (except mason.nvim) disabled, and the issue still occurs.

**Comment by chrisgrieser** [comment]
It seems other people are experiencing it as well. (I can confirm that I also have leftover emmylua_processes, so it does sound like the same bug.) https://www.reddit.com/r/neovim/comments/1mdtr4g/comment/n64bwvh/

**Comment by CppCXY** [comment]
I have confirmed that this is a deadlock caused by handling the configuration request.

**Comment by CppCXY** [comment]
you can download from https://github.com/EmmyLuaLs/emmylua-analyzer-rust/actions/runs/16703970387
Try to see if it is fixed.

**Comment by clason** [comment]
Looks to be fixed for me, thank you!

**Comment by chrisgrieser** [comment]
can confirm, it appears to be working now. Thank you for the quick fix!

**Comment by clason** [comment]
This issue can be closed then?
